galley/pkg/runtime/state.go

```diff
@@ -0,0 +1,154 @@
+
+package runtime
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+
+	"istio.io/api/config/mcp/v1alpha1"
+	"istio.io/istio/galley/pkg/mcp/snapshot"
+	"istio.io/istio/galley/pkg/runtime/resource"
+)
+
+// State is the in-memory state of Galley.
+type State struct {
+	schema *resource.Schema
+
+	// version counter is a nonce that generates unique ids for each updated view of State.
+	versionCounter int64
+
+	// entries for per-kind State.
+	entries     map[resource.Kind]*kindState
+	entriesLock sync.Mutex
+}
+
+// per-kind State.
+type kindState struct {
+	// The version number for the current State of the object. Every time entries or versions change,
+	// the version number also change
+	version  int64
+	entries  map[string]*v1alpha1.Envelope
+	versions map[string]resource.Version
+}
+
+func newState(schema *resource.Schema) *State {
+	return &State{
+		schema:  schema,
+		entries: make(map[resource.Kind]*kindState),
+	}
+}
+
+func (s *State) apply(event resource.Event) bool {
+	if _, ok := s.schema.LookupByKind(event.ID.Kind); !ok {
+		scope.Errorf("Received an source event for unknown kind: %v", event)
+		return false
+	}
+
+	pks := s.getKindState(event.ID.Kind)
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the version has changed.
+		if curVersion := pks.versions[event.ID.FullName]; curVersion == event.ID.Version {
+			scope.Debugf("Received event for the current, known version: %v", event)
+			return false
+		}
+
+		// TODO: Check for content-wise equality
+
+		entry, ok := s.envelopeResource(event)
+		if !ok {
+			return false
+		}
+
+		pks.entries[event.ID.FullName] = entry
+		pks.versions[event.ID.FullName] = event.ID.Version
+
+	case resource.Deleted:
+		delete(pks.entries, event.ID.FullName)
+		delete(pks.versions, event.ID.FullName)
+
+	default:
+		scope.Errorf("Unknown event kind: %v", event.Kind)
+		return false
+	}
+
+	s.versionCounter++
+	pks.version = s.versionCounter
+
+	return true
+}
+
+func (s *State) getKindState(kind resource.Kind) *kindState {
+	s.entriesLock.Lock()
+	defer s.entriesLock.Unlock()
+
+	pks, found := s.entries[kind]
+	if !found {
+		pks = &kindState{
+			entries:  make(map[string]*v1alpha1.Envelope),
+			versions: make(map[string]resource.Version),
+		}
+		s.entries[kind] = pks
+	}
+
+	return pks
+}
+
+func (s *State) buildSnapshot() snapshot.Snapshot {
+	s.entriesLock.Lock()
+	defer s.entriesLock.Unlock()
+
+	sn := snapshot.NewInMemory()
+
+	for kind, state := range s.entries {
+		entries := make([]*v1alpha1.Envelope, 0, len(state.entries))
+		for _, entry := range state.entries {
+			entries = append(entries, entry)
+		}
+
+		version := fmt.Sprintf("%d", state.version)
+		sn.Set(string(kind), version, entries)
+	}
+
+	return sn
+}
+
+func (s *State) envelopeResource(event resource.Event) (*v1alpha1.Envelope, bool) {
+	info, _ := s.schema.LookupByKind(event.ID.Kind)
+
+	serialized, err := proto.Marshal(event.Item)
```

Comment:
Use `types.MarshalAny()` here, e.g.

```golang
resource, err := types.MarshalAny(event.Item)
if err != nil { 
    nil, false 
}
entry := &v1alpha1.Envelope{
    Metadata: &v1alpha1.Metadata{
        Name: event.ID.FullName,
    },
    Resource: resource,
}

```

Reply:
That defacto hardwires googleapis URL.

Reply:
That seems reasonable. `google.protobuf.Any` implies googleapis URL. MCP (xDS) protocol also explicitly calls out use of `types.googleapis.com`. It's not clear what value we get by generalizing here.

Reply:
As mentioned before, I want to avoid incorporating implicit mappings & policy. This maybe turn out to be a mute concern, but additional 3 lines of code here for this purpose doesn't seem an excessive loss.

Reply:
Per offline discussion, let's address this issue holistically across galley (see https://github.com/istio/istio/issues/6437).